cli/tests/test_parallelize_command.rs

```diff
@@ -0,0 +1,400 @@
+    ├─┬─╮
+    │ │ ◉  48ecaa663297 3
+    │ │ │ ◉  9591d4874ae3 child-2a
+    ╭─┬───┤
+    │ │ │ ◉  5b36783cd11c a
+    │ │ ├─╯
+    │ ◉ │  e12f80116b44 2
+    │ ├─╯
+    ◉ │  7086b0bcbc06 1
+    ├─╯
+    ◉  000000000000
+    "###);
+}
+
+#[test]
+fn test_parallelize_failure_disjoint_target_commits() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let workspace_path = test_env.env_root().join("repo");
+
+    for n in 1..6 {
+        test_env.jj_cmd_ok(&workspace_path, &["commit", &format!("-m{n}")]);
+    }
+    test_env.jj_cmd_ok(&workspace_path, &["describe", "-m=6"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  b911505e443e 6
+    ◉  2e00cb15c7b6 5
+    ◉  9df3c87db1a2 4
+    ◉  9f5b59fa4622 3
+    ◉  d826910d21fb 2
+    ◉  dc0e5d6135ce 1
+    ◉  000000000000
+    "###);
+
+    insta::assert_snapshot!(test_env.jj_cmd_failure(&workspace_path, &["parallelize", "dc0", "9f5"]),@r###"
+    Error: Cannot parallelize since the target revisions are not connected.
+    "###);
+}
+
+#[test]
+fn test_parallelize_head_is_a_merge() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let workspace_path = test_env.env_root().join("repo");
+    test_env.jj_cmd_ok(&workspace_path, &["commit", "-m=1"]);
+    test_env.jj_cmd_ok(&workspace_path, &["commit", "-m=2"]);
+    test_env.jj_cmd_ok(&workspace_path, &["new", "root()"]);
+    test_env.jj_cmd_ok(&workspace_path, &["commit", "-m=a"]);
+    test_env.jj_cmd_ok(&workspace_path, &["commit", "-m=b"]);
+    test_env.jj_cmd_ok(
+        &workspace_path,
+        &["new", "description(2)", "description(b)", "-m=merged-head"],
+    );
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @    1a8db14a8cf0 merged-head
+    ├─╮
+    │ ◉  401e43e9461f b
+    │ ◉  66ea2ab19a70 a
+    ◉ │  d826910d21fb 2
+    ◉ │  dc0e5d6135ce 1
+    ├─╯
+    ◉  000000000000
+    "###);
+
+    insta::assert_snapshot!(test_env.jj_cmd_failure(&workspace_path,&["parallelize", "description(1)::"]),
+        @r###"
+    Error: Only the root commit is allowed to have parents which are not being parallelized.
+    Commit 1a8db14a8cf0 violates this rule.
+    "###);
+}
+
+#[test]
+fn test_parallelize_interior_target_is_a_merge() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let workspace_path = test_env.env_root().join("repo");
+    test_env.jj_cmd_ok(&workspace_path, &["describe", "-m=1"]);
+    test_env.jj_cmd_ok(&workspace_path, &["new", "root()", "-m=a"]);
+    test_env.jj_cmd_ok(
+        &workspace_path,
+        &["new", "description(1)", "description(a)", "-m=2"],
+    );
+    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=3"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
```

Comment:
There's one issue here. These commits actually cannot be parallelized at all because there is no single mutable root to include in the target revset.

I think I am going to change the rules such that multiple roots are allowed, but only if they have the same parents. This would allow us to parallelize `root()..` here, which is currently prohibited.

Reply:
This is fixed in an update I haven't pushed yet. In the update, multiple roots are allowed as long as they have the same parents.